runtime/compiler/optimizer/StringPeepholes.cpp

```diff
@@ -332,7 +379,11 @@ void TR_StringPeepholes::processBlock(TR::Block *block)
 
             TR::TreeTop *newTree = detectPattern(block, tt, true);
             if (newTree)
+               {
```

Comment:
I am concerned by the sequence which follows here. You are removing the OSR infrastructure before we have 'matched' the pattern and passed the point where we do the perform transformation to say that we are going to replace the peephole sequence with a new String of a private constructor targeting the specific sequence of appends we have found. This would be much more correct and much nicer if we skipped the OSR nodes as part of the pattern matching, ran the performTransformation and then did the remove of the OSR points before we did the manipulation to switch to the custom constructor. This pattern could see OSR infrastructure removed needlessly and could mean that doing a lastOptSubIndex in StringPeepholes could have unintended consequences on the OSR points and transitions seen by the OSR infrastructure which runs later.

Reply:
`detectPattern()` both matches the pattern and, if it finds a match, transforms the code - see the end of the function [starting here](https://github.com/eclipse/openj9/blob/7ebb4d4bef739fde5b4fcd6085f15449ffb25048/runtime/compiler/optimizer/StringPeepholes.cpp#L1863-L1890). I expect the other `detect*Pattern()` methods work similarly, considering that there doesn't seem to be any further transformation after skipping ahead to `newTree` in any of `processBlock`, `process`, or `perform` (except an attempt to inline some calls in the last of these).

Reply:
Yes. all the `detect*Pattern()` methods detect and transform the code at the same time, then return the last tree in the pattern.